bookwyrm/utils/tar.py

```diff
@@ -0,0 +1,46 @@
+"""manage tar files for user exports"""
+import io
+import tarfile
+from typing import Any
+from uuid import uuid4
+from django.core.files import File
+
+
+class BookwyrmTarFile(tarfile.TarFile):
+    """Create tar files for user exports"""
+
+    def write_bytes(self, data: bytes) -> None:
+        """Add a file containing bytes to the archive"""
+        buffer = io.BytesIO(data)
+        info = tarfile.TarInfo("archive.json")
+        info.size = len(data)
+        self.addfile(info, fileobj=buffer)
+
+    def add_image(self, image: Any, filename: str = None, directory: Any = "") -> None:
+        """
+        Add an image to the tar archive
+        :param str filename: overrides the file name set by image
+        :param str directory: the directory in the archive to put the image
+        """
+        if filename is not None:
+            file_type = image.name.rsplit(".", maxsplit=1)[-1]
+            filename = f"{directory}{filename}.{file_type}"
+        else:
+            filename = f"{directory}{image.name}"
+
+        info = tarfile.TarInfo(name=filename)
+        info.size = image.size
+
+        self.addfile(info, fileobj=image)
+
+    def read(self, filename: str) -> Any:
+        """read data from the tar"""
+        with self.extractfile(filename) as reader:
```

Comment:
> `error: Item "None" of "IO[bytes] | None" has no attribute "__enter__"`

```suggestion
        if reader := self.extractfile(filename):
```

(`extractfile` can return None, plus if it doesn't, it's just a regular object that doesn't need enter/exit with `with`)